snuba/state/stateful_count.py

```diff
@@ -0,0 +1,60 @@
+from threading import Thread
+from time import time
+from typing import Any, Sequence, Tuple
+
+from arroyo.processing.strategies.dead_letter_queue import CountInvalidMessagePolicy
+from arroyo.processing.strategies.dead_letter_queue.policies.abstract import (
+    InvalidMessage,
+)
+from arroyo.types import TPayload
+
+from snuba.redis import redis_client
+
+
+class StatefulCountInvalidMessagePolicy(CountInvalidMessagePolicy[TPayload]):
+    """
+    An extension of the CountInvalidMessagePolicy which is able to save and load
+    the state of counted hits in Redis
+    """
+
+    def __init__(self, redis_hash_name: str, limit: int, seconds: int = 60) -> None:
+        self.__name = redis_hash_name
+        self.__seconds = seconds
+        super().__init__(limit, seconds, self._load_state())
+
+    def handle_invalid_message(self, e: InvalidMessage) -> None:
+        """
+        Asynchronously updates Redis hash with a new hit while
+        the base Policy handles the invalid message
+        """
+        Thread(target=self._add_to_redis).start()
+        super().handle_invalid_message(e)
+
+    def _add_to_redis(self) -> None:
+        """
+        Increments the current time entry in Redis by 1
+        """
+        now = int(time())
+        p = redis_client.pipeline()
+        p.hincrby(self.__name, str(now), 1)
+        p.hkeys(self.__name)
+        self._prune(now, p.execute()[1])
+
+    def _prune(self, now: int, timestamps: Sequence[Any]) -> None:
```

Comment:
are the timestamps here also `int`s?

Reply:
Technically anything Redis returns is a `bytes` type, so it'd look something like `b'123'` which could be casted back into an `int`, but it's not necessary since those bytes can be compared with each other.

Reply:
If it's a `bytes` type, why not annotate it as `bytes` instead of `Any`. I would probably still cast it back into int to avoid unexpected things like the example Evan gave.